scheduler volumebinding: leverage PreFilterResult for bound local PVs

yibozhuang opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This change will leverage the new PreFilterResult
to reduce down the list of eligible nodes for pod
using Bound Local PVs during PreFilter stage so
that only the node(s) which local PV node affinity
matches will be considered in subsequent scheduling
stages.

Today, the NodeAffinity check is done during Filter
which means all nodes will be considered even though
there may be a large number of nodes that are not
eligible due to not matching the pod's bound local
PV(s)' node affinity requirement. Here we can
reduce down the node list in PreFilter to ensure that
during Filter we are only considering the reduced
list and thus can provide a more clear message to
users when node(s) are not available for scheduling
since the list only contains relevant nodes.

If error is encountered (e.g. PV cache read error) or
if node list reduction cannot be done (e.g. pod uses
no local PVs), then we will still proceed to consider
all nodes for the rest of scheduling stages.

Signed-off-by: Yibo Zhuang <yibzhuang@gmail.com>


### Testing example:
- 100 nodes cluster
- all 100 nodes tainted with different key, value
- pod has a bound PVC to local PV with node affinity on node `fake-0`

__without__ `PreFilter` change, scheduling output

```
  Type     Reason            Age   From               Message
  ----     ------            ----  ----               -------
  Warning  FailedScheduling  3s    default-scheduler  0/100 nodes are available: 1 node(s) had untolerated taint {key-fake-0: value-fake-0}, 1 node(s) had untolerated taint {key-fake-10: value-fake-10}, 1 node(s) had untolerated taint {key-fake-11: value-fake-11}, 1 node(s) had untolerated taint {key-fake-12: value-fake-12}, 1 node(s) had untolerated taint {key-fake-13: value-fake-13}, 1 node(s) had untolerated taint {key-fake-14: value-fake-14}, 1 node(s) had untolerated taint {key-fake-15: value-fake-15}, 1 node(s) had untolerated taint {key-fake-16: value-fake-16}, 1 node(s) had untolerated taint {key-fake-17: value-fake-17}, 1 node(s) had untolerated taint {key-fake-18: value-fake-18}, 1 node(s) had untolerated taint {key-fake-19: value-fake-19}, 1 node(s) had untolerated taint {key-fake-1: value-fake-1}, 1 node(s) had untolerated taint {key-fake-20: value-fake-20}, 1 node(s) had untolerated taint {key-fake-21: value-fake-21}, 1 node(s) had untolerated taint {key-fake-22: value-fake-22}, 1 node(s) had untolerated taint {key-fake-23: value-fake-23}, 1 nod ...
```

__with__ `PreFilter` change, scheduling output

```
  Type     Reason            Age   From               Message
  ----     ------            ----  ----               -------
  Warning  FailedScheduling  1s    default-scheduler  0/100 nodes are available: 1 node(s) had untolerated taint {key-fake-0: value-fake-0}. preemption: 0/100 nodes are available: 1 Preemption is not helpful for scheduling, 99 No preemption victims found for incoming pod.
```

Related to #109476

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #109876

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
scheduler volumebinding: leverage PreFilterResult to reduce down to only eligible node(s) for pod with bound claim(s) to local PersistentVolume(s)
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

yibozhuang commented:
/assign @Huang-Wei @cofyc @msau42

Huang-Wei commented on pkg/volume/util/util.go:
```diff
@@ -571,6 +571,34 @@ func IsLocalEphemeralVolume(volume v1.Volume) bool {
+// and return the node name for the PV node affinity if it is a local PV.
+// Local PV has node affinity in the form of:
+// nodeAffinity:
+//	required:
+//	  nodeSelectorTerms:
+//	  - matchExpressions:
+//		- key: kubernetes.io/hostname
+//		  operator: In
+//		  values:
+//		  - <node>
```
nit: use 2 spaces instead of tab for indenting purpose.

Huang-Wei commented on pkg/volume/util/util.go:
```diff
@@ -571,6 +571,34 @@ func IsLocalEphemeralVolume(volume v1.Volume) bool {
+			if len(nodeSelector.MatchExpressions) == 1 {
+				matchExpr := nodeSelector.MatchExpressions[0]
+				if matchExpr.Key == v1.LabelHostname && matchExpr.Operator == v1.NodeSelectorOpIn {
+					if len(matchExpr.Values) == 1 {
+						return matchExpr.Values[0], true
+					}
```
I'm thinking of relax the value check, so basically returns all legal host names instead of a single one? (and probably also relax the `len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms)` check)

@cofyc could you confirm this is the most common pattern to specify local PV?

yibozhuang replied:
AFAIK, local PV that is using the static provisioner have PV node affinity created like:
https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner/blob/master/pkg/discovery/discovery.go#L198

which is a single match expressions for key `kubernetes.io/hostname` with operation `In` and value the node name.

@msau42 @jsafrane can chime in if needed. Will bring this up with sig-storage.

cofyc replied:
as for local PVs provisioned by local-static-provisioner, there is only one hostname bound to PV

> I'm thinking of relax the value check, so basically returns all legal host names instead of a single one?

but +1 for this in scheduling, we can add all values under `kubernetes.io/hostname` key in potential nodes

Huang-Wei commented on pkg/volume/util/util.go:
```diff
@@ -571,6 +571,34 @@ func IsLocalEphemeralVolume(volume v1.Volume) bool {
+			}
+		}
+	}
+	return "", false
```
do we need 2 parameters?

Huang-Wei commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -356,6 +366,40 @@ func (b *volumeBinder) FindPodVolumes(pod *v1.Pod, boundClaims, claimsToBind []*
+	for _, pvc := range boundClaims {
+		pvName := pvc.Spec.VolumeName
+		pv, err := b.pvCache.GetPV(pvName)
+		if err != nil {
+			if _, ok := err.(*errNotFound); ok {
+				err = nil
+			}
+			errs = append(errs, err)
+			continue
+		}
```
```go
if err != nil && <err is errNotFound> {
    errs = append(errs, err)
    continue
}
```

yibozhuang replied:
Thanks, fixed.

Huang-Wei commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -356,6 +366,40 @@ func (b *volumeBinder) FindPodVolumes(pod *v1.Pod, boundClaims, claimsToBind []*
+
+	klog.V(4).InfoS("GetEligibleNodes: reduced down eligible nodes", "nodes", eligibleNodes, "error", errs)
+
+	if len(errs) > 0 {
+		return sets.NewString(), errors.NewAggregate(errs)
+	}
```
it means you're returning an empty string once there is an error, so why not return earlier when the first error is caught?

yibozhuang replied:
There is a logging line above and I think it would be helpful to log all the errors encountered for all Bound claims rather than see an issue and immediately return.

Huang-Wei commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -185,8 +185,17 @@ func (pl *VolumeBinding) PreFilter(ctx context.Context, state *framework.CycleSt
+	// Attempt to reduce down the number of nodes to consider in subsequent scheduling stages if pod has bound claims.
+	var result *framework.PreFilterResult
+	nodes, err := pl.Binder.GetEligibleNodes(boundClaims)
+	if err == nil && len(nodes) > 0 {
```
Given you don't rely on the `err`, GetEligibleNodes() can just have one return parameter (i.e., nodeNames).

yibozhuang replied:
Thanks, done.

Huang-Wei commented on pkg/volume/util/util.go:
```diff
@@ -571,6 +571,34 @@ func IsLocalEphemeralVolume(volume v1.Volume) bool {
+//		  values:
+//		  - <node>
+func GetLocalLocalPersistentVolumeNodeName(pv *v1.PersistentVolume) (string, bool) {
+	if pv != nil && pv.Spec.NodeAffinity != nil && pv.Spec.NodeAffinity.Required != nil {
```
stylish comment: the 5 ifs make the code intended too much and hard to read. I'd like code it like:

```go
if pv == nil || pv.Spec.NodeAffinity == nil || pv.Spec.NodeAffinity.Required == nil {
    return
}
if len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) != 1 {
    return
}
......

```

yibozhuang replied:
Sure thing, restructured to use negative checks and return early style.

Huang-Wei reviewed (commented):
will review tests in the 2nd round review.

Huang-Wei commented:
/ok-to-test

yibozhuang commented:
/retest

yibozhuang commented:
/retest

ddebroy commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

yibozhuang commented on pkg/volume/util/util.go:
```diff
@@ -571,6 +571,39 @@ func IsLocalEphemeralVolume(volume v1.Volume) bool {
+//         operator: In
+//         values:
+//         - <node>
+func GetLocalLocalPersistentVolumeNodeName(pv *v1.PersistentVolume) string {
```
Thanks, done.

yibozhuang commented:
/retest

msau42 commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -356,6 +366,38 @@ func (b *volumeBinder) FindPodVolumes(pod *v1.Pod, boundClaims, claimsToBind []*
+		// if the PersistentVolume has node affinity matching specific node(s), add them to the eligible nodes
+		nodeNames := util.GetPersistentVolumeNodeNames(pv)
+		for _, node := range nodeNames {
+			eligibleNodes.Insert(node)
```
Do we potentially want to take a union of the nodes across all pvs? So if pv1 is node1 and pv2 is node2, then it should return 0 nodes?

msau42 replied:
Building on this, should we distinguish between 0 nodes and all nodes?

yibozhuang replied:
Thanks @msau42 - I've updated the `GetEligibleNodes()` to:

1. return a bool `canReduceNodeList` along with the eligibleNodes set to indicate whether node reduction has been done, which will only be true if boundClaims bound to at least one local PV. If error occurred during get PV from cache or if no local PV, then this is false and all nodes will be considered

2. Using intersection for the eligibleNodes so for the case of PV1 to node1 and PV2 to node2, then intersection of those will be empty list

msau42 commented on pkg/volume/util/util.go:
```diff
@@ -571,6 +571,38 @@ func IsLocalEphemeralVolume(volume v1.Volume) bool {
+		return nil
+	}
+	matchExpr := nodeSelector.MatchExpressions[0]
+	if matchExpr.Key != v1.LabelHostname || matchExpr.Operator != v1.NodeSelectorOpIn {
```
So we want this to be an optimization only for local volumes? Maybe I would call this function `GetLocalPersistentVolumeNodeNames` instead.

yibozhuang replied:
Done, renamed to `GetLocalPersistentVolumeNodeNames`.

msau42 commented on pkg/scheduler/framework/plugins/volumebinding/binder_test.go:
```diff
@@ -2318,3 +2330,86 @@ func TestCapacity(t *testing.T) {
+		eligibleNodes sets.String
+	}
+
+	scenarios := map[string]scenarioType{
```
Can you add a scenario where there is a local pv and non-local pv?

yibozhuang replied:
Done, added test scenario, thanks.

yibozhuang commented:
/remove-lifecycle stale

msau42 commented:
@yibozhuang just wanted to check if you are still working on this

yibozhuang commented:
@msau42 yes I still want to get this merged. I am on vacation this week but should be able to get this PR rebased and addressing all feedback by early next week. If it doesn't make the cut for 1.26 then it's fine to get it merged post code-freeze for 1.27. Thanks for checking in on this.

yibozhuang commented:
Hi @msau42 @Huang-Wei, this PR is ready for another look, thanks

Huang-Wei reviewed (commented):
Overall the approach looks good. Some comments below:

yibozhuang commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -356,6 +370,61 @@ func (b *volumeBinder) FindPodVolumes(pod *v1.Pod, boundClaims, claimsToBind []*
+// existing bound claims or false then reduction cannot be made and all nodes should be considered.
+func (b *volumeBinder) GetEligibleNodes(boundClaims []*v1.PersistentVolumeClaim) (canReduceNodeList bool, eligibleNodes sets.String) {
+	canReduceNodeList = false
+	eligibleNodes = sets.NewString()
```
thanks, done.

yibozhuang commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -356,6 +370,61 @@ func (b *volumeBinder) FindPodVolumes(pod *v1.Pod, boundClaims, claimsToBind []*
+// in subsequent scheduling stages.
+//
+// The canReduceNodeList boolean is used to indicate if such eligible node reduction is applicable from the
+// existing bound claims or false then reduction cannot be made and all nodes should be considered.
```
thanks, reworded with the updated return of single nil or concrete set from feedback below.

yibozhuang commented on pkg/volume/util/util.go:
```diff
@@ -576,6 +576,39 @@ func IsLocalEphemeralVolume(volume v1.Volume) bool {
+	nodeSelector := pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0]
+	if len(nodeSelector.MatchExpressions) != 1 {
+		return nil
+	}
```
thanks, done.

yibozhuang commented on pkg/volume/util/util.go:
```diff
@@ -576,6 +576,39 @@ func IsLocalEphemeralVolume(volume v1.Volume) bool {
+	if len(nodeSelector.MatchExpressions) != 1 {
+		return nil
+	}
+	matchExpr := nodeSelector.MatchExpressions[0]
```
Thanks, changed this to iterate over all `MatchExpressions`
```go
nodeSelector := pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0]
for _, matchExpr := range nodeSelector.MatchExpressions {
        if matchExpr.Key == v1.LabelHostname && matchExpr.Operator == v1.NodeSelectorOpIn {
	        return matchExpr.Values
        }
}
return nil
```

Assume single match of `kubernetes.io/hostname in [...]` so return the first match expression found.

yibozhuang commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -356,6 +370,61 @@ func (b *volumeBinder) FindPodVolumes(pod *v1.Pod, boundClaims, claimsToBind []*
+		// if the PersistentVolume is local and has node affinity matching specific node(s),
+		// add them to the eligible nodes
+		nodeNames := util.GetLocalPersistentVolumeNodeNames(pv)
+		if nodeNames != nil {
```
thanks, fixed.

yibozhuang commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -356,6 +370,61 @@ func (b *volumeBinder) FindPodVolumes(pod *v1.Pod, boundClaims, claimsToBind []*
+		return false, sets.NewString()
+	}
+
+	if canReduceNodeList {
```
Yep agreed, changed the function to
```
GetEligibleNodes(boundClaims []*v1.PersistentVolumeClaim) (eligibleNodes sets.String)
```
where returning `nil` indicates node reduction cannot be done and all nodes are eligible.

yibozhuang commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -356,6 +370,61 @@ func (b *volumeBinder) FindPodVolumes(pod *v1.Pod, boundClaims, claimsToBind []*
+		return
+	}
+
+	notFound := &errNotFound{}
```
thanks, actually, thinking about this a bit more, we should bail if any error occurred here. If `GetPV()` returns PV not found, then we cannot determine whether we can make the node reduction, so we should just add to the overall error list and continue. I've made the change addressing this and `errNotFound` should be considered the same as other errors.

Huang-Wei commented on pkg/volume/util/util.go:
```diff
@@ -576,6 +576,38 @@ func IsLocalEphemeralVolume(volume v1.Volume) bool {
+	if pv == nil || pv.Spec.NodeAffinity == nil || pv.Spec.NodeAffinity.Required == nil {
+		return nil
+	}
+	if len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) != 1 {
+		return nil
+	}
+
+	nodeSelector := pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0]
```
I rethink the logic a bit. Given NodeSelectorTerms are ORed:

```go
type NodeSelector struct {
    // Required. A list of node selector terms. The terms are ORed.
    NodeSelectorTerms []NodeSelectorTerm
}
```

and NodeSelectorTerm.MatchExpressions is ANDed. We should do something like:

```go
nodes := sets.String()
for each nodeSelectorTerm:
    for each matchExpression:
        if it matches {key: hostname, operator: In}:
            // calculate an intersected node set
    // at the end of each inner loop, union the calculated node set to the final `nodes`
```

yibozhuang replied:
Thanks, done.

Huang-Wei commented on pkg/volume/util/util_test.go:
```diff
@@ -942,3 +943,249 @@ func TestGetPodVolumeNames(t *testing.T) {
+				Spec: v1.PersistentVolumeSpec{
+					NodeAffinity: &v1.VolumeNodeAffinity{
+						Required: &v1.NodeSelector{
+							NodeSelectorTerms: []v1.NodeSelectorTerm{
```
Construct an additional case for nodeAffinity like:

```
- NodeSelectorTerms[0]
    - MatchExpressions[0]: hostname in [node1, node2]
    - MatchExpressions[0]: hostname in [node2, node3]
- NodeSelectorTerms[1]:
    - MatchExpressions[0]: hostname in [node1]
```

The final result should be [node1, node2]

yibozhuang replied:
Thanks, added test case described above.

Huang-Wei reviewed (commented):
Some comment on the matching semantics of NodeSelectorTerms and MatchExpressions. Otherwise looks good.

yibozhuang committed:
scheduler volumebinding: leverage PreFilterResult

This change will leverage the new PreFilterResult
to reduce down the list of eligible nodes for pod
using Bound Local PVs during PreFilter stage so
that only the node(s) which local PV node affinity
matches will be cosnidered in subsequent scheduling
stages.

Today, the NodeAffinity check is done during Filter
which means all nodes will be considered even though
there may be a large number of nodes that are not
eligible due to not matching the pod's bound local
PV(s)' node affinity requirement. Here we can
reduce down the node list in PreFilter to ensure that
during Filter we are only considering the reduced
list and thus can provide a more clear message to
users when node(s) are not available for scheduling
since the list only contains relevant nodes.

If error is encountered (e.g. PV cache read error) or
if node list reduction cannot be done (e.g. pod uses
no local PVs), then we will still proceed to consider
all nodes for the rest of scheduling stages.

Signed-off-by: Yibo Zhuang <yibzhuang@gmail.com>

yibozhuang commented:
/retest

Huang-Wei commented:
/approve for scheduling

Leave review for the storage part to @msau42 .

yibozhuang commented:
@msau42 will you have cycles to do another review on this? Thanks in advance

dims commented:
If you still need this PR then please rebase, if not, please close the PR

msau42 commented:
sorry for the delay, will try to take a look soon

msau42 commented:
/lgtm
/approve

Thanks!
